cocotbext/uart/uart.py

```diff
@@ -120,11 +123,25 @@ def clear(self):
     async def wait(self):
         await self._idle.wait()
 
+    async def add_glitch(self, data, time, unit):
+        if self._glitch_generator is not None:
+            mask = next(self._glitch_generator)
+            if any(mask):
+                bit_val = data.value.integer
+                for k in mask:
+                    if bit_val == 1:
+                        k = 1 if k == 0 else 0
+                    data.value = k
+                    await Timer(int(time/len(mask)),unit)
+
+    async def wait_bit(self, data, time, unit):
+        cocotb.start_soon(self.add_glitch(data,time,unit))
+        await Timer(time, unit)
+
     async def _run(self, data, baud, bits, stop_bits):
         self.active = False
 
-        bit_t = Timer(int(1e9/self.baud), 'ns')
-        stop_bit_t = Timer(int(1e9/self.baud*stop_bits), 'ns')
+        bit_t = lambda: self.wait_bit(data, int(1e9/self.baud), 'ns')
```

Comment:
The whole point of "caching" the events is to not continuously create and destroy event objects.  Presumably you're making this change to support changing the baud rate dynamically?  If so, let's make baud and stop_bits properties, store the events as instance parameters, and swap out the events when these properties are changed.

Reply:
Also, it might make sense to rework all of the time math in terms of sim steps, which avoids passing units around all over the place.